opentelemetry/src/main/java/io/grpc/opentelemetry/OpenTelemetryTracingModule.java

```diff
@@ -0,0 +1,408 @@
+    }
+
+    @Override
+    public void streamClosed(io.grpc.Status status) {
+      endSpanWithStatus(span, status);
+    }
+  }
+
+  private final class ServerTracer extends ServerStreamTracer {
+    private final Span span;
+    volatile int streamClosed;
+    private int seqNo;
+
+    ServerTracer(String fullMethodName, @Nullable Span remoteSpan) {
+      checkNotNull(fullMethodName, "fullMethodName");
+      this.span =
+          otelTracer.spanBuilder(generateTraceSpanName(true, fullMethodName))
+              .setParent(remoteSpan == null ? null : Context.current().with(remoteSpan))
+              .startSpan();
+    }
+
+    /**
+     * Record a finished stream and mark the current time as the end time.
+     *
+     * <p>Can be called from any thread without synchronization.  Calling it the second time or more
+     * is a no-op.
+     */
+    @Override
+    public void streamClosed(io.grpc.Status status) {
+      if (streamClosedUpdater != null) {
+        if (streamClosedUpdater.getAndSet(this, 1) != 0) {
+          return;
+        }
+      } else {
+        if (streamClosed != 0) {
+          return;
+        }
+        streamClosed = 1;
+      }
+      endSpanWithStatus(span, status);
+    }
+
+    @Override
+    public void outboundMessageSent(
+        int seqNo, long optionalWireSize, long optionalUncompressedSize) {
+      recordOutboundMessageSentEvent(span, seqNo, optionalWireSize, optionalUncompressedSize);
+    }
+
+    @Override
+    public void inboundMessageRead(
+        int seqNo, long optionalWireSize, long optionalUncompressedSize) {
+      if (optionalWireSize != optionalUncompressedSize) {
+        recordInboundCompressedMessage(span, seqNo, optionalWireSize);
+      }
+    }
+
+    @Override
+    public void inboundMessage(int seqNo) {
+      this.seqNo = seqNo;
+    }
+
+    @Override
+    public void inboundUncompressedSize(long bytes) {
+      recordInboundMessageSize(span, seqNo, bytes);
+    }
+  }
+
+  @VisibleForTesting
+  final class ServerTracerFactory extends ServerStreamTracer.Factory {
+    @SuppressWarnings("ReferenceEquality")
+    @Override
+    public ServerStreamTracer newServerStreamTracer(String fullMethodName, Metadata headers) {
+      Context context = contextPropagators.getTextMapPropagator().extract(
+          Context.current(), headers, metadataGetter
+      );
+      Span remoteSpan = Span.fromContext(context);
+      if (remoteSpan == Span.getInvalid()) {
+        remoteSpan = null;
+      }
+      return new ServerTracer(fullMethodName, remoteSpan);
+    }
+  }
+
+  @VisibleForTesting
+  final class TracingClientInterceptor implements ClientInterceptor {
+
+    @Override
+    public <ReqT, RespT> ClientCall<ReqT, RespT> interceptCall(
+        MethodDescriptor<ReqT, RespT> method, CallOptions callOptions, Channel next) {
+      Span clientSpan = otelTracer.spanBuilder(
+          generateTraceSpanName(false, method.getFullMethodName()))
+          .startSpan();
```

Comment:
How can we specify the parent span for this if one exists?

Reply:
Great question! Opentelemetry sdk by default pulls in the current span on the io.opentelemetry.context as it's parent span. 
https://github.com/open-telemetry/opentelemetry-java/blob/main/sdk/trace/src/main/java/io/opentelemetry/sdk/trace/SdkSpanBuilder.java#L185-L186
Users will just makeCurrent() of the parent span using opentelemetry context API.
Interceptor are run in channel thread.